From b554ab5c9bba537b4a282b3863ead6b3b52dd2e1 Mon Sep 17 00:00:00 2001
From: Jakub Radtke <jakub.radtke@intel.com>
Date: Fri, 15 Jan 2021 00:46:55 -0500
Subject: [PATCH 120/136] Enable bitmap support for external metadata

The patch enables the implementation of a write-intent bitmap for external
metadata.
Configuration of the internal bitmaps for non-native metadata requires the
extension in superswitch to perform an additional sysfs setup before the
array is activated.

Signed-off-by: Jakub Radtke <jakub.radtke@intel.com>
Signed-off-by: Jes Sorensen <jsorensen@fb.com>
---
 Assemble.c |  7 +++++++
 Create.c   | 12 ++++++++++++
 mdadm.h    |  3 +++
 3 files changed, 22 insertions(+)

diff --git a/Assemble.c b/Assemble.c
index 48556d8c..5c6aca92 100644
--- a/Assemble.c
+++ b/Assemble.c
@@ -2094,6 +2094,13 @@ int assemble_container_content(struct supertype *st, int mdfd,
 		}
 	}
 
+	/*
+	 * Before activating the array, perform extra steps required
+	 * to configure the internal write-intent bitmap.
+	 */
+	if (content->consistency_policy == CONSISTENCY_POLICY_BITMAP &&
+	    st->ss->set_bitmap)
+		st->ss->set_bitmap(st, content);
 
 	if (start_reshape) {
 		int spare = content->array.raid_disks + expansion;
diff --git a/Create.c b/Create.c
index 51f8826b..b50cb246 100644
--- a/Create.c
+++ b/Create.c
@@ -989,6 +989,18 @@ int Create(struct supertype *st, char *mddev,
 				st->ss->free_super(st);
 				goto abort_locked;
 			}
+			/*
+			 * Before activating the array, perform extra steps
+			 * required to configure the internal write-intent
+			 * bitmap.
+			 */
+			if (info_new.consistency_policy ==
+				    CONSISTENCY_POLICY_BITMAP &&
+			    st->ss->set_bitmap &&
+			    st->ss->set_bitmap(st, &info)) {
+				st->ss->free_super(st);
+				goto abort_locked;
+			}
 
 			/* update parent container uuid */
 			if (me) {
diff --git a/mdadm.h b/mdadm.h
index 60575af0..273fa509 100644
--- a/mdadm.h
+++ b/mdadm.h
@@ -1007,6 +1007,9 @@ extern struct superswitch {
 	int (*add_internal_bitmap)(struct supertype *st, int *chunkp,
 				   int delay, int write_behind,
 				   unsigned long long size, int may_change, int major);
+	/* Perform additional setup required to activate a bitmap.
+	 */
+	int (*set_bitmap)(struct supertype *st, struct mdinfo *info);
 	/* Seek 'fd' to start of write-intent-bitmap.  Must be an
 	 * md-native format bitmap
 	 */
-- 
2.25.1

